### PR TITLE
Add "How it Works" section

### PR DESCRIPTION
**Pull request by msporny**
This PR adds a section called "How it Works" to the Data Integrity specification to conceptually describe how the technology described by the specification works.

There are a few images in the PR, so I'm attaching a screen shot for how the rendered diagram looks:

image

At present, the section is a bit out of place, and probably belongs more in the Introduction than in this section. I'll do an Editorial move later once we've agreed to the content.

***
<a href="https://pr-preview.s3.amazonaws.com/w3c/vc-data-integrity/pull/47.html" title="Last updated on Sep 4, 2022, 3:57 PM UTC (bf77220)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/vc-data-integrity/47/fb108b9...bf77220.html" title="Last updated on Sep 4, 2022, 3:57 PM UTC (bf77220)">Diff</a>

### CODE REVIEW

File: index.html [hunk 1]
```diff
@@ -325,8 +325,8 @@ <h3>How it Works</h3>
 Proof'.
                ">
           <figcaption style="text-align: center;">
-To create a cryptographic proof data is transformed, hashed, and
-cryptographically protected.
+To verify a cryptographic proof data is transformed, hashed, and
```
Comment (TallTed) [inline index.html]:
```suggestion
To verify a cryptographic proof, data is transformed, hashed, and
```

File: index.html [hunk 2]
```diff
@@ -260,6 +260,96 @@ <h3>Design Goals and Rationale</h3>
...
+The operation of Data Integrity is conceptually simple. To create a
+cryptographic proof, the following steps are performed: 1) Transformation,
+2) Hashing, and 3) Proof Generation.
+        </p>
+
+        <figure id="hiw-creation">
+          <img style="margin: auto; display: block; width: 100%;"
+               src="diagrams/hiw-creation.svg" alt="
+Diagram showing the three steps involved in the creation of a cryptographic
+proof. The diagram is layed out left to right with a blue box labeled 'Data'
```
Comment (TallTed) [inline index.html]:
```suggestion
proof. The diagram is laid out left to right with a blue box labeled 'Data'
```

File: index.html [hunk 3]
```diff
@@ -260,6 +260,96 @@ <h3>Design Goals and Rationale</h3>
...
+on the very left. The blue box travels, left to right, through three subsequent
+yellow arrows labeled 'Transform Data', 'Hash Data', and 'Generate Proof'. The
+resulting blue box at the far right is labeled 'Data with Proof'.
+               ">
+          <figcaption style="text-align: center;">
+To create a cryptographic proof data is transformed, hashed, and
```
Comment (TallTed) [inline index.html]:
```suggestion
To create a cryptographic proof, data is transformed, hashed, and
```

File: index.html [hunk 4]
```diff
@@ -260,6 +260,96 @@ <h3>Design Goals and Rationale</h3>
...
+<a href="https://en.wikipedia.org/wiki/Binary-to-text_encoding">
+binary-to-text</a> encoding. <dfn>Hashing</dfn> is a process that calculates an
+identifier for the transformed data using a
+<a href="https://en.wikipedia.org/wiki/Cryptographic_hash_function">
+cryptographic hash function</a>. Possible cryptographic hash functions
+include
+<a href="https://en.wikipedia.org/wiki/Cryptographic_hash_function#SHA-3">
+SHA-3</a> and
+<a href="https://en.wikipedia.org/wiki/Cryptographic_hash_function#BLAKE3">
+BLAKE-3</a>. <dfn class="lint-ignore">Proof Generation</dfn> is a process that
```
Comment (TallTed) [inline index.html]:
```suggestion
BLAKE-3</a>.
        </p>

        <p>
<dfn class="lint-ignore">Proof Generation</dfn> is a process that
```

File: index.html [hunk 5]
```diff
@@ -260,6 +260,96 @@ <h3>Design Goals and Rationale</h3>
...
+          </figcaption>
+        </figure>
+
+        <p>
+<dfn>Transformation</dfn> is a process that takes input data and prepares it for
+the hashing process. Possible transformations include
+<a href="https://en.wikipedia.org/wiki/Canonicalization">canonicalization</a>
+and
+<a href="https://en.wikipedia.org/wiki/Binary-to-text_encoding">
+binary-to-text</a> encoding. <dfn>Hashing</dfn> is a process that calculates an
```
Comment (TallTed) [inline index.html]:
```suggestion
binary-to-text</a> encoding.
        </p>

        <p>
<dfn>Hashing</dfn> is a process that calculates an
```

File: index.html [hunk 6]
```diff
@@ -260,6 +260,96 @@ <h3>Design Goals and Rationale</h3>
...
+<a href="https://en.wikipedia.org/wiki/Cryptographic_hash_function">
+cryptographic hash function</a>. Possible cryptographic hash functions
+include
+<a href="https://en.wikipedia.org/wiki/Cryptographic_hash_function#SHA-3">
+SHA-3</a> and
+<a href="https://en.wikipedia.org/wiki/Cryptographic_hash_function#BLAKE3">
+BLAKE-3</a>. <dfn class="lint-ignore">Proof Generation</dfn> is a process that
+calculates a value that protects the integrity of the input data from
+modification or otherwise proves a certain desired threshold of trust. This
+process is conceptually similar to how a wax seal can be used on an envelope
```
Comment (TallTed) [inline index.html]:
```suggestion
process is conceptually similar to the way a wax seal can be used on an envelope
```

File: index.html [hunk 7]
```diff
@@ -260,6 +260,96 @@ <h3>Design Goals and Rationale</h3>
...
+containing a letter to establish trust in the sender and that the letter has not
```
Comment (TallTed) [inline index.html]:
```suggestion
containing a letter to establish trust in the sender and show that the letter has not
```

File: index.html [hunk 8]
```diff
@@ -260,6 +260,96 @@ <h3>Design Goals and Rationale</h3>
...
+        <p>
+To verify a cryptographic proof, the following steps are performed:
+1) Transformation, 2) Hashing, and 3) Proof Verification.
+        </p>
+
+        <figure id="hiw-verification">
+          <img style="margin: auto; display: block; width: 100%;"
+               src="diagrams/hiw-verification.svg" alt="
+Diagram showing the three steps involved in the verification of a cryptographic
+proof. The diagram is layed out left to right with a blue box labeled
```
Comment (TallTed) [inline index.html]:
```suggestion
proof. The diagram is laid out left to right with a blue box labeled
```

File: index.html [hunk 9]
```diff
@@ -260,6 +260,96 @@ <h3>Design Goals and Rationale</h3>
...
+'Data with Proof' on the very left. The blue box travels, left to right, through
```
Comment (TallTed) [inline index.html]:
```suggestion
'Data with Proof' on the far left. The blue box travels, left to right, through
```

File: index.html [hunk 10]
```diff
@@ -260,6 +260,96 @@ <h3>Design Goals and Rationale</h3>
...
+cryptographic proof, the following steps are performed: 1) Transformation,
+2) Hashing, and 3) Proof Generation.
+        </p>
+
+        <figure id="hiw-creation">
+          <img style="margin: auto; display: block; width: 100%;"
+               src="diagrams/hiw-creation.svg" alt="
+Diagram showing the three steps involved in the creation of a cryptographic
+proof. The diagram is layed out left to right with a blue box labeled 'Data'
+on the very left. The blue box travels, left to right, through three subsequent
```
Comment (TallTed) [inline index.html]:
```suggestion
on the far left. The blue box travels, left to right, through three subsequent
```

File: index.html [hunk 11]
```diff
@@ -260,6 +260,96 @@ <h3>Design Goals and Rationale</h3>
...
+               ">
+          <figcaption style="text-align: center;">
+To verify a cryptographic proof data is transformed, hashed, and
+checked for correctness.
+          </figcaption>
+        </figure>
+
+        <p>
+During verification, the <a>transformation</a> and <a>hashing</a> steps are
+conceptually the same as described above. <dfn class="lint-ignore">Proof
```
Comment (TallTed) [inline index.html]:
```suggestion
conceptually the same as described above.
        </p>

        <p>
<dfn class="lint-ignore">Proof
```

File: index.html [hunk 12]
```diff
@@ -260,6 +260,96 @@ <h3>Design Goals and Rationale</h3>
...
+Verification</dfn> is a process that checks to see if the input data can be
+trusted by applying a cryptographic proof verification function. Possible
```
Comment (TallTed) [inline index.html]:
```suggestion
Verification</dfn> is a process that applies a cryptographic proof
verification function to see if the input data can be trusted. Possible
```

File: index.html [hunk 13]
```diff
@@ -260,6 +260,96 @@ <h3>Design Goals and Rationale</h3>
...
+<a href="https://en.wikipedia.org/wiki/Digital_signature">digital signatures</a>
+and
+<a href="https://en.wikipedia.org/wiki/Proof_of_stake">proofs of stake</a>.
+        </p>
+
+        <p>
+This specification details how cryptographic software architects and
+implementers can package these processes together into things called
+<a>cryptographic suites</a> and provide them to application developers for the
+purposes of protecting the integrity of their data in transit and at rest.
```
Comment (TallTed) [inline index.html]:
```suggestion
purposes of protecting the integrity of application data in transit and at rest.
```

**Review by TallTed — CHANGES_REQUESTED** [review]
Mostly minor punctuation and grammar.

**Review by thereisnogabe — APPROVED** [review]
nice work, this is easy to read and comprehensive

**Review by dlongley — APPROVED** [review]
I think this is a good start. We'll want to make sure that the description here also cleanly maps to what happens with BBS signatures, for example. In BBS signatures, transformation occurs, but the transformed data produces N messages that are hashed and signed, i.e., there are multiple signatures. This still seems to map to me but we'll have to see if others agree, ideally from the BBS suite spec writers.

**Comment by msporny** [comment]
> In BBS signatures, transformation occurs, but the transformed data produces N messages that are hashed and signed, i.e., there are multiple signatures.

Yes, I winced in the same way when I re-read the text yesterday. I do think we'll need to do some minor modifications to handle the "hash N messages" case. I'll try to put that in a revision before merging.

**Review by iherman — APPROVED** [review]
My comment is purely stylistic and, I admit, picky... (I approve the text regardless of taking this comment or not)

The sentence:

> This process is conceptually similar to...

is trying to provide an analogy to a reader who is unfamiliar with the concept of digital signatures and/or proofs of stake. And that is fine. However, wouldn't it be better, for consistency's sake, to add a similar analogy for what a hash function means? (I guess we can get away without an analogy for the transformation bit.)

**Review by mprorock — APPROVED** [review]
awesome improvement!

**Comment by msporny** [comment]
> However, wouldn't it be better, for consistency's sake, to add a similar analogy for what a hash function means? (I guess we can get away without an analogy for the transformation bit.)

I have added this here: 0c83e1ccbfe665a612752ab8c62ee25eff58a8c0

**Comment by msporny** [comment]
Editorial, multiple reviews, changes requested and made, no objections, merging.

**Pull request merged**